### PR TITLE
Tpos signature hf

### PR DESCRIPTION
**Pull request by durkmurder**

### CODE REVIEW

File: src/blocksigner.cpp [hunk 1]
```diff
@@ -45,15 +46,18 @@ bool CBlockSigner::SignBlock()
 
         if(refBlock.IsTPoSBlock())
         {
-            CKeyID merchantKeyID;
-            if(!refContract.merchantAddress.GetKeyID(merchantKeyID))
+            if(!refContract.scriptMerchantAddress.IsPayToPublicKeyHash()) {
```
Comment (AlexITC) [inline src/blocksigner.cpp:49]: Would you mind elaborating why is this required?

File: src/blocksigner.cpp [hunk 2]
```diff
@@ -69,8 +73,13 @@ bool CBlockSigner::SignBlock()
             scriptType = GetScriptTypeFromDestination(destination);
         }
     }
-//?
-    return CHashSigner::SignHash(refBlock.IsTPoSBlock() ? refBlock.GetTPoSHash() : refBlock.GetHash(), keySecret, scriptType, refBlock.vchBlockSig);
+
+    const auto &hash = refBlock.IsTPoSBlock() ? refBlock.GetTPoSHash() : refBlock.GetHash();
+    if (nChainHeight >= Params().GetConsensus().nTPoSSignatureUpgradeHFHeight) {
```
Comment (AlexITC) [inline src/blocksigner.cpp:78]: I'd suggest to create a common function that has tests to check this, instead of keep writing the same check everywhere.

Or, even better, just write a helper thing that handles signing contracts, and verifying signatures.

File: src/blocksigner.cpp [hunk 3]
```diff
@@ -85,24 +94,26 @@ bool CBlockSigner::CheckBlockSignature() const
...
     }
-    else
-    {
+    else {
         return true;
     }
 
     std::string strError;
-    return CHashSigner::VerifyHash(hashMessage, destination, refBlock.vchBlockSig, strError);
+    if (nChainHeight >= Params().GetConsensus().nTPoSSignatureUpgradeHFHeight) {
```
Comment (AlexITC) [inline src/blocksigner.cpp:114]: Same duplication stated above.

File: src/chainparams.cpp [hunk 4]
```diff
@@ -99,6 +99,7 @@ class CMainParams : public CChainParams {
         consensus.nPosTargetSpacing = 1 * 60; // XSN: 1 minutes
         consensus.nPosTargetTimespan = 60 * 40;
         consensus.nPoSUpdgradeHFHeight = 898488; // 4 December 2019
+        consensus.nTPoSSignatureUpgradeHFHeight = 1264447;
```
Comment (AlexITC) [inline src/chainparams.cpp:102]: It's working commenting what's this for, and why.

File: src/chainparams.cpp [hunk 5]
```diff
@@ -422,13 +423,13 @@ class CRegTestParams : public CChainParams {
             0
         };
 
-        base58Prefixes[PUBKEY_ADDRESS] = std::vector<unsigned char>(1,140);
-        base58Prefixes[SCRIPT_ADDRESS] = std::vector<unsigned char>(1,19);
-        base58Prefixes[SECRET_KEY] =     std::vector<unsigned char>(1,239);
-        base58Prefixes[EXT_PUBLIC_KEY] = {0x04, 0x35, 0x87, 0xCF};
-        base58Prefixes[EXT_SECRET_KEY] = {0x04, 0x35, 0x83, 0x94};
+        base58Prefixes[PUBKEY_ADDRESS] = std::vector<unsigned char>(1,76);
```
Comment (AlexITC) [inline src/chainparams.cpp:426]: Are these changes actually needed? It was useful for testing but is it ok to keep the change? Also, I'd add the details to the README for using the regtest.

File: src/masternode-payments.cpp [hunk 6]
```diff
@@ -1115,8 +1115,11 @@ void AdjustMasternodePayment(CMutableTransaction &tx, const CTxOut &txoutMastern
         if(tposContract.IsValid()) // here we have 3 outputs, first as stake reward, second as tpos reward, third as MN reward
         {
             masternodePayment /= 100; // to calculate percentage
-            tx.vout[i - 1].nValue -= masternodePayment * tposContract.stakePercentage; // adjust reward for owner.
-            tx.vout[i].nValue -= masternodePayment * (100 - tposContract.stakePercentage); // adjust reward for merchant
+            tx.vout[i - 1].nValue -= masternodePayment * (100 - tposContract.nOperatorReward); // adjust reward for owner.
```
Comment (AlexITC) [inline src/masternode-payments.cpp:1118]: I'd create a common function for this, and make sure it has proper tests, a mistake would be very bad for us

File: src/tpos/tposutils.cpp [hunk 7]
```diff
@@ -96,213 +91,202 @@ bool TPoSUtils::IsTPoSMerchantContract(CWallet *wallet, const CTransactionRef &t
...
-        {
-            strError = strprintf("%s : TPoS contract signature is invalid %s", __func__, strError);
+
+        CTxDestination tposAddress;
+        if(!ExtractDestination(tmpContract.scriptTPoSAddress, tposAddress)) {
+            strError = strprintf("%s : TPoS contract invalid tpos address", __func__);
             return error(strError.c_str());
         }
+
+        if (nBlockHeight >= Params().GetConsensus().nTPoSSignatureUpgradeHFHeight) {
```
Comment (AlexITC) [inline src/tpos/tposutils.cpp:275]: You can get this logic in another reusable function

File: src/tpos/tposutils.cpp [hunk 8]
```diff
@@ -374,105 +360,250 @@ bool TPoSUtils::IsMerchantPaymentValid(CValidationState &state, const CBlock &bl
     }
 
     CKeyID coinstakeKeyID;
-    if(!merchantAddress.GetKeyID(coinstakeKeyID))
-        return state.DoS(0, error("IsMerchantPaymentValid -- ERROR: coin stake was paid to invalid address\n"),
-                         REJECT_INVALID, "bad-merchant-payee", true);
+    // legacy way supports only P2PKH, after HF support both P2PKH and P2WPKH
+    if (nBlockHeight >= Params().GetConsensus().nTPoSSignatureUpgradeHFHeight) {
```
Comment (AlexITC) [inline src/tpos/tposutils.cpp:364]: A reusable function could help to not keep checking the height everywhere

File: src/tpos/tposutils.cpp [hunk 9]
```diff
@@ -374,105 +360,250 @@ bool TPoSUtils::IsMerchantPaymentValid(CValidationState &state, const CBlock &bl
...
+                                    TPoSContract contract;
+                            ds >> contract;
+                            contract.txContract = tx;
+                            if (contract.IsValid()) {
+                                return contract;
+                            }
+                        } else if (tokens.size() > 3 && tokens.size() < 6) {
+
+                            CBitcoinAddress tposAddress(ParseAddressFromMetadata(tokens[1]));
+                            CBitcoinAddress merchantAddress(ParseAddressFromMetadata(tokens[2]));
```
Comment (AlexITC) [inline src/tpos/tposutils.cpp:581]: > merchant address is not P2PKH
I see you are validating this in another file, shouldn't you do the same to prevent creating such invalid contracts?

File: src/tpos/tposutils.cpp [hunk 10]
```diff
@@ -374,105 +360,250 @@ bool TPoSUtils::IsMerchantPaymentValid(CValidationState &state, const CBlock &bl
...
+                            }
+
+                            // legacy contract
+                            TPoSContract contract(tx,
+                                                  merchantAddress.Get(),
+                                                  tposAddress.Get(),
+                                                  commission, vchSignature);
+                            contract.nVersion = 1; // legacy version
+
+                            if(contract.IsValid() && commission > 0 && commission < 100) {
```
Comment (AlexITC) [inline src/tpos/tposutils.cpp:596]: These checks could go in a common function to avoid getting confused on whether you accept 0 or not.

File: src/validation.cpp [hunk 11]
```diff
@@ -613,6 +613,14 @@ static bool AcceptToMemoryPoolWorker(const CChainParams& chainparams, CTxMemPool
         return state.Invalid(false, REJECT_DUPLICATE, "txn-already-in-mempool");
     }
 
+    if (TPoSUtils::IsTPoSContract(ptx)) {
+        TPoSContract contract;
+        std::string strError;
+        if (!TPoSUtils::CheckContract(ptx, contract, chainActive.Height(), true, false, strError)) {
+            return state.Invalid(false, REJECT_INVALID, "tpos-invaid-contract");
```
Comment (AlexITC) [inline src/validation.cpp:620]: inva**l**id

File: src/version.h [hunk 12]
```diff
@@ -9,7 +9,9 @@
  * network protocol versioning
  */
 
-static const int PROTOCOL_VERSION = 70209;
+static const int PROTOCOL_VERSION = 70210;
+
+static const int PRETPOS_SIGNTURE_FIX_PROTO_VERSION = 70209;
```
Comment (AlexITC) [inline src/version.h:14]: It's worth adding a comment.

File: src/wallet/wallet.cpp [hunk 13]
```diff
@@ -475,9 +475,9 @@ void CWallet::FillCoinStakePayments(CMutableTransaction &transaction,
 
     unsigned int percentage = 100;
 
-    if(tposContract.IsValid())
-        percentage = tposContract.stakePercentage; // first vout is owner
-
+    if(tposContract.IsValid()) {
+        percentage = (100 - tposContract.nOperatorReward); // first vout is owner
```
Comment (AlexITC) [inline src/wallet/wallet.cpp:479]: A reusable function would help.

File: src/blocksigner.cpp (hunk 1 above)
Comment (durkmurder) [reply src/blocksigner.cpp:49]: Currently we accept merchant addresses only in P2PKH format, so we avoid signing blocks that would get rejected

File: src/chainparams.cpp (hunk 4 above)
Comment (durkmurder) [reply src/chainparams.cpp:102]: we will set the date later

File: src/chainparams.cpp (hunk 5 above)
Comment (durkmurder) [reply src/chainparams.cpp:426]: Yeah, should be ok, doesn't look too bad

File: src/masternode-payments.cpp (hunk 6 above)
Comment (durkmurder) [reply src/masternode-payments.cpp:1118]: true
